easybuild/tools/module_generator.py

```diff
@@ -260,7 +268,18 @@ def prepend_paths(self, key, paths, allow_abs=False, expand_relpaths=True):
         paths = self._filter_paths(key, paths)
         if paths is None:
             return ''
-        return self.update_paths(key, paths, prepend=True, allow_abs=allow_abs, expand_relpaths=expand_relpaths)
+        if key == 'PYTHONPATH':
+            # Special condition for PYTHONPATHs that match the standard pattern,
+            # replace with EBPYTHONPREFIX which is added to python sys path at runtime
+            ret = ''
+            for path in paths:
+                if re.match(r'lib/python\d+\.\d+/site-packages', path):
+                    self.log.info("Replaced PYTHONPATH %s with EBPYTHONPREFIX", path)
+                    ret += self._update_paths('EBPYTHONPREFIX', [''], prepend, allow_abs, expand_relpaths)
+                else:
+                    ret += self._update_paths(key, [path], prepend, allow_abs, expand_relpaths)
+            return ret
```

Comment:
How about: 
```suggestion
            # Special condition for PYTHONPATHs that match the standard pattern,
            # replace with EBPYTHONPREFIX which is added to python sys path at runtime
            python_paths = {path for path in paths if re.match(r'lib/python\d+\.\d+/site-packages', path)}
            if python_paths:
                self.log.info("Replaced PYTHONPATH %s with EBPYTHONPREFIX", python_paths)
                paths = [path for path in paths if path not in python_paths]
                ret = self._update_paths('EBPYTHONPREFIX', [''], prepend, allow_abs, expand_relpaths)
                if paths
                    ret += self._update_paths(key, paths, prepend, allow_abs, expand_relpaths)
                return ret
```

This avoids adding `$EBPYTHONPREFIX` potentially multiple times and does only a single call to `_update_paths` for the remainder if any

Reply:
Sure, though that scenario would only actually occur if there was multiple difference python versions matching (e.g. python2.7 + python3.11 etc.), which would have been very broken using PYTHONPATH already, so `python_paths` should never have more than 1 element.

Reply:
> so python_paths should never have more than 1 element.

That is true. Might be worth to check for and show a meaningful error message